### PR TITLE
Extend `ci-link` to PR

### LINKED ISSUE #4086: Show `ci-link` for PR next to it's title/number
Similarly to repo CI status:
https://github.com/sindresorhus/refined-github/blob/main/source/features/ci-link.tsx
image

I propose adding a CI status next to PR name/title/number.
Reasons:
1. On long PRs - you need to scroll A LOT to see CI status ([for example](https://github.com/sindresorhus/refined-github/pull/4023) )
2. When you're in another tab - like changes - you need to switch tabs to know CI status. So you need to switch tab to conversations + scroll to bottom.
3. When you're already scrolled in changes view - you need to *lose context* to know CI status. Like half way thru review, or [direct link](https://github.com/sindresorhus/refined-github/pull/3810/files#diff-5a831ea67cf5cf8703b0de46901ab25bd191f56b320053be9332d9a3b0d01d15R224)

### PR DESCRIPTION
**Pull request by kidonng**
Resolve #4086

Meta unchanged though

## Test URLs

#4098

## Screenshot

image

image

### CODE REVIEW

**Comment by kidonng** [comment]
Dumb me 😅

File: source/features/sync-pr-commit-title.tsx [hunk 1]
```diff
@@ -16,10 +17,6 @@ function getCommitTitleField(): HTMLInputElement | undefined {
 	return select<HTMLInputElement>(`${mergeFormSelector} #merge_title_field`);
 }
 
-function getPRNumber(): string {
```
Comment (yakov116) [inline source/features/sync-pr-commit-title.tsx:19]: I think you should just export it from here.

Also this is not domReady safe

Comment (kidonng) [reply source/features/sync-pr-commit-title.tsx:19]: > I think you should just export it from here.

Well we may need this in the future. It doesn't hurt.

> this is not domReady safe

What!? This is in the HTML!

image

Comment (yakov116) [reply source/features/sync-pr-commit-title.tsx:19]: use `getConversationNumber()`

File: source/features/ci-link.tsx [hunk 2]
```diff
@@ -5,31 +5,51 @@ import * as pageDetect from 'github-url-detection';
...
 ));
 
-async function init(): Promise<false | void> {
-	const icon = await getIcon() as HTMLElement | undefined;
+const getPRIcon = onetime(async () => fetchDom(
+	buildRepoURL('pull', getConversationNumber()!, 'commits'), '.js-commits-list-item:last-of-type .commit-build-statuses'
+));
+
+async function initRepo(): Promise<false | void> {
+	const icon = await getRepoIcon() as HTMLElement | undefined;
```
Comment (fregante) [inline source/features/ci-link.tsx:23]: Let's try to move these types to the `get*Icon` functions so we avoid assertions

File: source/features/ci-link.tsx [hunk 3]
```diff
@@ -5,31 +5,51 @@ import * as pageDetect from 'github-url-detection';
...
+	}
+
+	icon.classList.add('rgh-ci-link');
+	if (onetime.callCount(getPRIcon) > 1) {
+		icon.style.animation = 'none';
+	}
+
+	// Append to PR title and floating header
+	select('.gh-header-title')!.append(icon);
+	select('.gh-header-number')!.after(icon.cloneNode(true));
```
Comment (fregante) [inline source/features/ci-link.tsx:50]: Do the icons stay up there even after new comments, label changes, title changes?

File: source/features/sync-pr-commit-title.tsx (hunk 1 above)
Comment (fregante) [reply source/features/sync-pr-commit-title.tsx]: @yakov116 I'm confused, why do we have this function at all? `sync-pr-commit-title` runs on PRs too, so the number is in the URL

Comment (yakov116) [reply source/features/sync-pr-commit-title.tsx]: That one needed the `#` so someone added it. I noticed it a while ago, but just left it.

Comment (fregante) [reply source/features/sync-pr-commit-title.tsx]: @yakov116 you stopped the blame too early. That line of code is ancient 😃 https://github.com/sindresorhus/refined-github/commit/ed2d06f0b0c441f16a7b444f3a8fc55e6f986924

File: source/features/ci-link.tsx (hunk 3 above)
Comment (kidonng) [reply source/features/ci-link.tsx:50]: Nope... Guess we need selector-observer

File: source/features/ci-link.tsx [hunk 4]
```diff
@@ -1,35 +1,67 @@
...
 }
 
+async function initPR(): Promise<false | void> {
+	const icon = await getPRIcon();
+	if (!icon) {
+		return false;
+	}
+
+	icon.classList.add('rgh-ci-link');
+	if (onetime.callCount(getPRIcon) > 1) {
```
Comment (kidonng) [inline source/features/ci-link.tsx:45]: This doesn't work... I have some ideas but none of them looks good :(

File: source/features/sync-pr-commit-title.tsx (hunk 1 above)
Comment (yakov116) [reply source/features/sync-pr-commit-title.tsx]: Ok I'll excuse you this time 😉

File: source/features/ci-link.tsx (hunk 3 above)
Comment (yakov116) [reply source/features/ci-link.tsx]:
```js
	init() {
		observeElement(select('#partial-discussion-header')!.parentElement!, init);
	}
```


This should be enough no?

**Comment by fregante** [comment]
Position, size and alignment need to be reviewed in both cases
Screen Shot

Screen Shot 1

File: source/features/ci-link.tsx (hunk 3 above)
Comment (fregante) [reply source/features/ci-link.tsx]: Hopefully 👆

**Comment by fregante** [comment]
Try placing it in the `#1234` element so they always wrap together and we avoid orphans:

Screen Shot 3

**Comment by kidonng** [comment]
- Updated size to match with the text
- Grouped the icon with PR number

**Comment by fregante** [comment]
> * Updated size to match with the text

Looks good


Screen Shot 2021-04-03 at 22 40 39

Screen Shot 2021-04-03 at 22 40 45


> * Grouped the icon with PR number

Doesn't actually follow its purpose though

Screen Shot 2021-04-03 at 22 41 16


---

- [ ] Readme needs to be updated to mention it also works on PRs

**Comment by kidonng** [comment]
> Doesn't actually follow its purpose though

Works on my machine™️

https://user-images.githubusercontent.com/44045911/113847138-bc66b380-97c9-11eb-83f0-d33b07050eda.mov

File: source/features/ci-link.tsx [hunk 5]
```diff
@@ -1,35 +1,77 @@
...
 import * as pageDetect from 'github-url-detection';
 
 import features from '.';
 import fetchDom from '../helpers/fetch-dom';
-import {buildRepoURL} from '../github-helpers';
+import {buildRepoURL, getConversationNumber} from '../github-helpers';
 
 // Look for the CI icon in the latest 2 days of commits #2990
-const getIcon = onetime(async () => fetchDom(
+const getRepoIcon = onetime(async () => fetchDom<HTMLElement>(
```
[the remark this answers is not in the document]
Comment (kidonng) [reply source/features/ci-link.tsx]: This means we can't use `onetime(func).callCount` anymore

File: source/features/ci-link.tsx [hunk 6]
```diff
@@ -1,35 +1,74 @@
...
+	if (onetime.callCount(getRepoIcon) > 1) {
 		icon.style.animation = 'none';
 	}
 
 	// Append to title (aware of forks and private repos)
 	select('[itemprop="name"]')!.parentElement!.append(icon);
 }
 
+async function initPR(): Promise<false | void> {
+	const icon = await getPRIcon() as HTMLElement | undefined;
```
Comment (kidonng) [inline source/features/ci-link.tsx:44]: I can't inline this type in `getPRIcon` because eslint thinks `select` returns `HTMLElement` by default while typescript sees it `Element`.

File: source/features/ci-link.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (fregante) [reply source/features/ci-link.tsx:14]: I forgot about that

File: source/features/ci-link.tsx (hunk 6 above)
Comment (fregante) [reply source/features/ci-link.tsx]: Must be a TS bug. Fixed by using `<HTMLElement>` on the select itself

**Comment by fregante** [comment]
> Works on my machine™️ (Chrome 89.0.4389.114 on macOS)

Not For Me © 1969-2021 on Firefox (MPL) and Safari (Jony I.V.E.)

File: source/features/ci-link.tsx [hunk 7]
```diff
@@ -23,8 +23,9 @@ const getPRIcon = onetime(async () => {
 		document :
 		await fetchDom(buildRepoURL('pull', getConversationNumber()!, 'commits'));
 
-	// TS bug requires a generic here 🤷‍♂️
-	return select<HTMLElement>('.js-commits-list-item:last-of-type .commit-build-statuses', base);
+	// TS bug requires this 🤷‍♂️
```
Comment (yakov116) [inline source/features/ci-link.tsx:26]:
```suggestion
	// TS bug does not allow us to directly return this 🤷‍♂️
```

Maybe?

**Comment by kidonng** [comment]
> Not For Me © 1969-2021 on Firefox (MPL) and Safari (Jony I.V.E.)

I guess they treat flexbox differently. Try changing `.f1-light` from `span` to `div`, it wraps in Safari but not in Chrome.

**Comment by yakov116** [comment]
The X icon still needs some CSS help

image

File: source/features/ci-link.tsx [hunk 8]
```diff
@@ -22,7 +22,7 @@ const getPRIcon = onetime(async () => {
 		await fetchDom(buildRepoURL('pull', getConversationNumber()!, 'commits'));
 
 	// TS bug does not allow us to directly return this 🤷‍♂️
-	const icon = select('.js-commits-list-item:last-of-type .commit-build-statuses', base);
+	const icon = select.last('.js-commits-list-item:last-of-type .commit-build-statuses', base);
```
Comment (fregante) [inline source/features/ci-link.tsx:25]: Doesn't :last-of-type already do this? If not, is it needed anymore?

Comment (fregante) [reply source/features/ci-link.tsx:25]: Also does the new fetchDom type let us use `return select` now?

File: source/features/ci-link.tsx [hunk 9]
```diff
@@ -5,24 +5,57 @@ import * as pageDetect from 'github-url-detection';
...
+		return false;
+	}
+
+	icon.classList.add('rgh-ci-link', 'ml-2');
+	if (onetime.callCount(getPRIcon) > 1) {
+		icon.style.animation = 'none';
+	}
+
+	const headers = select.all(':is(.gh-header-title .f1-light, .js-sticky h1):not(.rgh-ci-link-heading)');
+
```
Comment (fregante) [inline source/features/ci-link.tsx:41]:
```suggestion
```

File: source/features/ci-link.tsx (hunk 8 above)
Comment (yakov116) [reply source/features/ci-link.tsx:25]: I think its still needed. (Can't check right now)

>The commits can span over a few days, we need the last one on the list

We need the last group and the last one on the list.

File: source/features/ci-link.tsx [hunk 10]
```diff
@@ -38,5 +70,14 @@ void features.add(__filebasename, {
...
-	init
+	init: initRepo
+}, {
+	include: [
+		pageDetect.isPR
+	],
+	additionalListeners: [
+		onConversationHeaderUpdate
+	],
+	awaitDomReady: pageDetect.isPRCommitList(),
```
Comment (yakov116) [inline source/features/ci-link.tsx:81]: TS help please

File: source/features/ci-link.tsx (hunk 8 above)
Comment (yakov116) [reply source/features/ci-link.tsx]: > Also does the new fetchDom type let us use `return select` now?

No it does not

File: source/features/ci-link.tsx (hunk 10 above)
Comment (fregante) [reply source/features/ci-link.tsx:81]: Can't be done. `pageDetect.isPRCommitList` would only be called on the first page load, but not on Ajax. You'll have to use `false` here and if `domReady` is needed conditionally, add `if (pageDetect.isPRCommitList()) await domLoaded` in the init function

**Review by zarnisoe2353 — APPROVED** [review]

**Review by zarnisoe2353 — COMMENTED** [review]
4100

**Review by zarnisoe2353 — COMMENTED** [review]
3826

**Comment by yakov116** [comment]
@zarnisoe2353 STOP

**Comment by yakov116** [comment]
This PR is moot until after https://github.com/sindresorhus/refined-github/issues/4260 is fixed

**Comment by yakov116** [comment]
@kidonng can you finish this up? I would love to see it merged?

**Comment by fregante** [comment]
Closing this PR for now. Feel free to create a new PR after addressing all the reviews and squashing the commits
